runtime/lua/vim/lsp/util.lua

```diff
@@ -1304,27 +1324,64 @@ function M.open_floating_preview(contents, syntax, opts)
     opts = { opts, 't', true };
   }
   opts = opts or {}
+  opts.fancy_markdown = opts.fancy_markdown == nil and true or opts.fancy_markdown
+  opts.close_events = opts.close_events or {"CursorMoved", "CursorMovedI", "BufHidden", "InsertCharPre"}
+
+  local bufnr = api.nvim_get_current_buf()
+
+  -- check if this popup is focusable and we need to focus
+  if opts.focus_id then
+    -- Go back to previous window if we are in a focusable one
+    local current_winnr = api.nvim_get_current_win()
+    if npcall(api.nvim_win_get_var, current_winnr, opts.focus_id) then
+      api.nvim_command("wincmd p")
+      return bufnr, current_winnr
+    end
+    do
+      local win = find_window_by_var(opts.focus_id, bufnr)
```

Comment:
I think there's a bug in the issue tracker where two windows can exist simultaneously from handlers, not sure if if focus_id and bufnr will currently always guarantee the correct buffer (it should)

Reply:
I'm not 100% sure I understand what you mean, but the bufnr should be correct I think, since `focus_id` is something like `hover` or `signature`.

If you set the cursor on a method that has both signature help and hover doc. Then pressing `hover`, `signature_help`, `hover` will correctly focus the hover foat.

Reply:
As in, focus_id is not actually a numerical (unique per windows) but categorical ID (unique per window type), and there is an open bug where multiple windows can be open at a time, not sure if it's possible for multiple windows with the same categorical ID to be open at a time

Reply:
that property is set on the floating window, not the parent window, so this should not be an issue?

Technically multiple floats can exists for the same focus_id, but becuase of the default close events, this won't happen.

If you'd change (or don't set) the events, then I believe it would still work as expected:
1. go to method one and `hover`
2. got to method two and `hover`
3. the open hover flloat will get the focus

The scenario above doesn't won't happen because of the default close events, but it could be changed.

Reply:
I'm aware, but if you're searching for focus_id's across all floating windows, and multiple floating windows have the same focus_id, I believe that will cause issues.

The issue in question is https://github.com/neovim/neovim/issues/11508

Reply:
yeah, it depends on how we want the popups to behave.

Right now there can be only one popup per focus_id open at one time.

Do we want to change this to never allow more than one lsp poup open? I can change it like that.

Reply:
Yeah, I think we shouldn't allow more than one popup per focus_id type for now (it's how it's supposed to be right now).

Reply:
Just to be sure, I don't make any changes then right? :)

> we shouldn't allow more than one popup per focus_id type for now

that's how it currently works. You can have multiple floats open, but at most one for every focus_id type. So never two or more `hover` floats for example.

Reply:
Yes, my point in bringing it up is there is currently an open bug report that implies that there is a corner case in which we can have multiple floating windows of the same type (hover, signature help, etc.) as well as multiple of different types. If with this PR, the former behavior is fixed (or at least, not readily reproducible), we shouldn't block on merging this, I just brought it up in review as this is possibly undefined behavior.